viewing_party/party.py

```diff
@@ -1,23 +1,248 @@
+
+def get_user_watched_titles(user_data):
+    #returns a set of titles user has watched.
+    user_watched_titles = []
+    for movie in user_data['watched']:
+        user_watched_titles.append(movie["title"])
+    return set(user_watched_titles)
+
+def get_friends_watched_titles(user_data):
+    #returns a set of titles of movies friend has watched.
+    friend_watched_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_titles.append(movie["title"])
+    return set(friend_watched_titles)
+
+def get_favorite_watched_titles(user_data):
+    #returns a set of titles user has watched.
+    user_watched_titles = []
+    for movie in user_data['favorites']:
+        user_watched_titles.append(movie["title"])
+    return set(user_watched_titles)
+
+
+def get_friends_unique_watched(user_data):
+    #this takes in user_data (with keys including 'watched' and 'friends')
+    # it returns a list of dictionaries of all the movies a friend(s) has watched that user hasn't.  
+    friend_watched_titles = get_friends_watched_titles(user_data)
+    user_watched_titles = get_user_watched_titles(user_data)
+    friends_unique_titles = friend_watched_titles - user_watched_titles
+
+    #retrieve info from user_data matching the titles
+    unique_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in friends_unique_titles and movie not in unique_watched:
+                unique_watched.append(movie)
+    return unique_watched
+
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    #this will return a list of movies that:
+    #the user has not watched
+    #at least one of the user's friends has watched
+    #is on a streaming service that the user has. 
+
+    friends_uniqued_watched = get_friends_unique_watched(user_data)
+    # here we filter for movies in the users subscription list.
+    recommendations = []
+    for movie in friends_uniqued_watched:
+        if movie["host"] in user_data["subscriptions"]:
+            recommendations.append(movie)
+
+    return recommendations
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    # this looks at the movies the user has watched.  
+    # it finds the most watched genre.
+    #look at unique watched by friends
+    #return what movies on that list are in the most watched genre.
+
+    top_genre = get_top_genre(user_data)
+    
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    recommendations_to_user = []
+
+    for movie in friends_unique_watched:
+        if movie["genre"] == top_genre:
+            recommendations_to_user.append(movie)
+
+    return recommendations_to_user

```

Comment:
Again, great use of helper functions. The breakdown really streamlines your code to be easily readable and organized.